umake/frameworks/ide.py

```diff
@@ -68,69 +68,66 @@ def __init__(self):
-                       "file=/technology/epp/downloads/release/luna/R/" \
-                       "eclipse-standard-luna-R-linux-gtk{arch}.tar.gz{suf}" \
-                       "&r=1"
 
     def __init__(self, category):
         super().__init__(name="Eclipse",
-                         description=_("Pure Eclipse Luna (4.4)"),
+                         description=_("Eclipse Java"),
                          category=category, only_on_archs=['i386', 'amd64'],
-                         download_page=None,
+                         download_page='https://www.eclipse.org/downloads/',
+                         checksum_type=ChecksumType.sha512,
                          dir_to_decompress_in_tarball='eclipse',
                          desktop_filename='eclipse.desktop',
                          required_files_path=["eclipse"],
                          packages_requirements=['openjdk-7-jdk'])
 
-    def download_provider_page(self):
-        """First, we need to fetch the MD5, then kick off the proceedings.
+        self.bits = '' if platform.machine() == 'i686' else 'x86_64'
 
-        This could actually be done in parallel, in a future version.
-        """
-        logger.debug("Preparing to download MD5.")
+    def download_sha512(self, sha512_url):
+        logger.debug("Preparing to Download SHA1")
 
-        arch = platform.machine()
-        if arch == 'i686':
-            md5_url = self.DOWNLOAD_URL_PAT.format(arch='', suf='.md5')
-        elif arch == 'x86_64':
-            md5_url = self.DOWNLOAD_URL_PAT.format(arch='-x86_64', suf='.md5')
-        else:
-            logger.error("Unsupported architecture: {}".format(arch))
-            UI.return_main_screen(status_code=1)
-
-        @MainLoop.in_mainloop_thread
         def done(download_result):
-            res = download_result[md5_url]
+            res = download_result[sha512_url]
 
             if res.error:
                 logger.error(res.error)
                 UI.return_main_screen(status_code=1)
 
             # Should be ASCII anyway.
-            md5 = res.buffer.getvalue().decode('utf-8').split()[0]
-            logger.debug("Downloaded MD5 is {}".format(md5))
-
-            logger.debug("Preparing to download the main archive.")
-            if arch == 'i686':
-                download_url = self.DOWNLOAD_URL_PAT.format(arch='', suf='')
-            elif arch == 'x86_64':
-                download_url = self.DOWNLOAD_URL_PAT.format(arch='-x86_64',
-                                                            suf='')
-            self.download_requests.append(DownloadItem(download_url, Checksum(ChecksumType.md5, md5)))
-            self.start_download_and_install()
+            self.sha512 = res.buffer.getvalue().decode('utf-8').split()[0]
+        DownloadCenter(urls=[DownloadItem(sha512_url, None)], on_done=done, download=False)
+        # It has to wait for the download and decode of the sha512sum
+        while self.sha512 is None:
+            pass
```

Comment:
Indeed, that's part of the issue.

The thing is that you can't give a direct list as you need to parse the page first to get this url result and launch the required download (which happens in another thread), and the while loop defeat that purpose.

So, the more elegant way is a little bit more involved, what I would do:
override (parent implementation in baseinstaller.py):
```
    @MainLoop.in_mainloop_thread
    def get_metadata_and_check_license(self, result):
```
In that one, strip the license checking (maybe this method would need to be renamed at some point ;)). result is the whole initial download page in a dictionary (listed by url). This is the code which is calling parse_download_link.
Then, when parse_download_link calls, your
```
 DownloadCenter(urls=[DownloadItem(sha512_url, None)], on_done=self.get_sha_and_start_download, download=False)
```
We know we have the download url already at that point, as the sha512 url is based from it.
So, you just need, in the new get_sha_and_start_download callback (you need to have it running in main thread, so, something around the following):

```
@MainLoop.in_mainloop_thread
def det_sha_and_start_download(self, download_result)
    res = download_result[sha512_url]
    sha512 = res.buffer.getvalue().decode('utf-8').split()[0]
    # you get and store self.download_url
    self.download_requests.append(DownloadItem(self.url, Checksum(self.checksum_type, sha512)))
    self.start_download_and_install()
```

That way, no while loop. You need to add of course empty or invalid conditions logging and error as in the initial get_metadata_and_check_license() method. As you know you have a checksum and no license, the method override can be quite simplified.

Does it make sense? I'm happy to maybe add an additional hook at some point when the download url assets are depending on each others, compared to be just in a fixed set of multiple download pages